server: make the Drain interfaces available to SQL-only servers

knz opened the pull request:
Fixes #74412

First commit from #76292; the reviewers are invited to ignore it in this PR.

See individual commits for details.

cc @cockroachdb/obs-inf-prs

knz committed:
server: move the drain delegate code to a separate function

Release note: None

knz committed:
server: split the shutdown code to a different function

Release note: None

knz committed:
server: split the local RPC handler to a separate function

Release note: None

knz committed:
server: lift the drain logic into a properly encapsulated object

Release note: None

knz committed:
server: make SQL-only server use a regular grpcServer

Release note: None

knz committed:
server: implement the Drain RPC for SQL-only servers

Release note: None

knz committed:
server: define DrainClients on both TestServer and TestTenant

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76294)
<!-- Reviewable:end -->

knz committed:
cli,server: make the CLI handle shutdown in a uniform manner

This brings the `mt start-sql` command "under the fold" with regards
to server shutdown. It now properly waits for client connections to go
away, subject to the same cluster settings as usual.

Release note: None

cameronnunez reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Good work!

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, 1 of 1 files at r5, 3 of 3 files at r6, 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-:-MvZSDuz0_KvNPH5K9lk:b-pv0wtg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cameronnunez and @knz)

___
*[pkg/server/tenant.go, line 78 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-MvZShFJALgM74QMnzck:-MvZShFJALgM74QMnzcl:b-97xtx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/20bdaf146680441e81c2b09c98d1c5c0ec124c51/pkg/server/tenant.go#L78)):*
> ```Go
> }
> 
> // SQLServerWrapper is a utility structs that encapsulates
> ```

nit: "is a utility struct"


<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/server/tenant_admin.go:
```diff
@@ -84,4 +99,52 @@ func (t *tenantAdminServer) Health(
+	ctx = t.AnnotateCtx(ctx)
+
+	// Which node is this request for?
+	parsedInstanceID, local, err := t.status.parseInstanceID(req.NodeId)
```
FYI: We do not track instance ID in the CC operator. The operator thinks in terms of pods. Determining the instance ID of the pod we want to remove is more difficult than sending an RPC to the specific pod we want to drain. I suspect we will pass NodeId == "" when sending a Drain rpc.

jeffswenson reviewed (approved):
LGTM

It would be nice if there were an rpc to cancel a drain. There is a mismatch between the CRDB draining state and the serverless draining state. When the operator decides to reduce the number of pods allocated to a serverless cluster, it marks the surplus pods in a draining state. The sqlproxy maintains existing connections to draining pods, but it will not send new connections to them. If the cluster's utilization increases, the pods are put back in a serving state. If there was a way to cancel CRDB drains, we could notify the sql server it is draining as soon as the operator decides to reduce capacity. With one way drains, the operator needs to wait to send the drain until it is ready to remove the pod.

knz reviewed (commented):
> It would be nice if there were an rpc to cancel a drain.

Point taken. let's discuss this in #76423.
I also encourage you to raise that issue to your nearest PM, so we can put it in our roadmap.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-:-MvcfB2J-AxQSANkKfBw:bkcugrg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cameronnunez and @JeffSwenson)

___
*[pkg/server/tenant.go, line 78 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-MvZShFJALgM74QMnzck:-MvcfqVM97IHegx_n2ah:b-5098c6) ([raw file](https://github.com/cockroachdb/cockroach/blob/20bdaf146680441e81c2b09c98d1c5c0ec124c51/pkg/server/tenant.go#L78)):*
<details><summary><i>Previously, cameronnunez (Cameron Nunez) wrote…</i></summary><blockquote>

nit: "is a utility struct"
</blockquote></details>

Fixed. Thanks
___
*[pkg/server/tenant_admin.go, line 126 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-MvU-3J4Vz2y4whodeqN-r6-126:-MvcfrX62QlHpg6TATGw:bnh6ueh) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec0757bf9bbcd6ebd64c06a76b9b6dbda93f7b0/pkg/server/tenant_admin.go#L126)):*
> I suspect we will pass NodeId == "" when sending a Drain rpc.

Yes, that works.

The semantics as implemented here ensure that we're transparently compatible with what is supported on storage nodes, so the command-line tooling can stay the same.


<!-- Sent from Reviewable.io -->

knz commented:
TFYRs!

bors r=cameronnunez,JeffSwenson

knz commented:
bors r-

knz commented:
bors r=cameronnunez,JeffSwenson

knz commented:
hm lint failures
bors r-

knz commented:
bors r=cameronnunez,JeffSwenson
